core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -380,29 +409,37 @@ DEFINE_UNARY_FUNCTION_EVAL(logb, 2);
 
 #undef DEFINE_UNARY_FUNCTION_EVAL
 
-#define DEFINE_BINARY_FUNCTION_EVAL(FUNC, ULP_FACTOR)                     \
-  struct MathBinaryFunction_##FUNC {                                      \
-    template <typename T, typename U>                                     \
-    static KOKKOS_FUNCTION auto eval(T x, U y) {                          \
-      static_assert(                                                      \
-          std::is_same<decltype(Kokkos::FUNC((T)0, (U)0)),                \
-                       math_binary_function_return_type_t<T, U>>::value); \
-      return Kokkos::FUNC(x, y);                                          \
-    }                                                                     \
-    template <typename T, typename U>                                     \
-    static auto eval_std(T x, U y) {                                      \
-      static_assert(                                                      \
-          std::is_same<decltype(std::FUNC((T)0, (U)0)),                   \
-                       math_binary_function_return_type_t<T, U>>::value); \
-      return std::FUNC(x, y);                                             \
-    }                                                                     \
-    static KOKKOS_FUNCTION double ulp_factor() { return ULP_FACTOR; }     \
-  };                                                                      \
-  using kk_##FUNC = MathBinaryFunction_##FUNC;                            \
-  template <>                                                             \
-  struct math_function_name<MathBinaryFunction_##FUNC> {                  \
-    static constexpr char name[] = #FUNC;                                 \
-  };                                                                      \
+#define DEFINE_BINARY_FUNCTION_EVAL(FUNC, ULP_FACTOR)                       \
+  struct MathBinaryFunction_##FUNC {                                        \
+    template <typename T, typename U>                                       \
+    static KOKKOS_FUNCTION auto eval(T x, U y) {                            \
+      static_assert(                                                        \
+          std::is_same<decltype(Kokkos::FUNC((T)0, (U)0)),                  \
+                       math_binary_function_return_type_t<T, U>>::value);   \
+      return Kokkos::FUNC(x, y);                                            \
+    }                                                                       \
+    template <typename T, typename U>                                       \
+    static auto eval_std(T x, U y) {                                        \
+      if constexpr (std::is_same<T, KE::half_t>::value ||                   \
+                    std::is_same<T, KE::bhalf_t>::value) {                  \
+        return std::FUNC(static_cast<float>(x), static_cast<int>(y));       \
```

Comment:
Why are we converting the2nd argument to int?  What does that even mean with `hypot` or `pow` for instance.
This looks wrong to me.

Reply:
We cannot write a implicit conversion constructor for converting half to float since we cannot assign precedence to operator overloads in floating_point_wrapper. That being said, this looks wrong to me too. When either T or U is half_t or bhalf_t, we should upcast both arguments to float.